---
title: "R Notebook"
output: html_notebook
---
Quick little side analysis- how related are these things by study? Is this an issue to be aware of?
```{r}
pacman::p_load(data.table, tidyr, dplyr, ggplot2, stringr,stats, cowplot, magrittr)
source("/work-zfs/abattle4/ashton/snp_networks/gwas_decomp_ldsc/src/plot_functions.R")
```


```{r}
udler_n_scaled <- fread("/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/full_signed_matrix.tsv")
udler_Beta <- fread("/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/processed_data/beta_signed_matrix.tsv")
udler_se <- fread("/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/processed_data/se_matrix.tsv")
udler_n <- fread("/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/processed_data/sample_counts_matrix.tsv") %>% print()
```
GEt cor, ploto with heatmaps
Issue with the N scaled data, need to lok into this....
make sure my inputs are valid...

```{r}
n_scaled_fix <- udler_Beta[,-1] / (udler_se[,-1] * sqrt(udler_n[,-1]))
z <- udler_Beta[,-1] / (udler_se[,-1])
n_scaled_fix$Type_2_Diabetes
plotCorrelationHeatmap(cor(n_scaled_fix[,-24])) #drop diabetes, its wierd
plotCorrelationHeatmap(cor(z[,-24]))
```
Why is the type2 diabaetes data so jacked up? I need to look back into this....
These do not appear to be any different....
Is there a cohort specific effect?
```{r}
sample.sizes <- fread("/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/file.names.withcohorts.april.txt") %>% filter(Inclusion == 1) %>% print()
library(reshape2)
image.df <- melt(cor(z[,-24])) %>% rename("Trait" = Var1) %>% left_join(., sample.sizes, by = "Trait") %>% select(Trait, Var2, value, Cohort) %>% rename("Trait1" = Trait, "Cohort1" = Cohort) %>% rename('Trait' = Var2) %>% left_join(., sample.sizes, by = "Trait") %>% rename("Trait2" = Trait, "Cohort2" = Cohort) %>% print()
```
What is the average by group?
Plain z-scores
```{r}
image.df %>% filter(Cohort1 == Cohort2) %>% group_by(Cohort1) %>% summarize("avg" = mean(value^2), "n" = n()) %>% filter(n > 1)
message("Overall avg")
print(mean(cor(z[,-24])^2))
```
By the corrected z-scores
```{r}
image.df.scaled <- melt(cor(n_scaled_fix[,-24])) %>% rename("Trait" = Var1) %>% left_join(., sample.sizes, by = "Trait") %>% select(Trait, Var2, value, Cohort) %>% rename("Trait1" = Trait, "Cohort1" = Cohort) %>% rename('Trait' = Var2) %>% left_join(., sample.sizes, by = "Trait") %>% rename("Trait2" = Trait, "Cohort2" = Cohort) %>% print()
image.df.scaled %>% filter(Cohort1 == Cohort2) %>% group_by(Cohort1) %>% summarize("avg" = mean(value^2), "n" = n()) %>% filter(n > 1)
message("Overall avg")
print(mean(cor(n_scaled_fix[,-24])^2))
```
Scaling by N the way they do doesn't deal with this problem. No suprises there...
Specify colors for plotting
```{r}
library(RColorBrewer)
col <- brewer.pal(12, "Paired")
colors <- sample.sizes %>% select(Cohort) %>% distinct() %>% 
  mutate("color" = c(col[1:3], col[3], col[4:12])) %>% print()
```


Yea, they are totally enriched for correlation. But this is unsurprising- these studies all look at related phenotypes, so I'd expect them to be highly correlated.
If we had 2 of the same traits from different cohorts in here, that would be helpful. Maybe that's what I need to include....
```{r}
  library(reshape2)
#My version, vanilla hclust
  cormat <- reorder_cormat(cor(n_scaled_fix[,-24]))
  melted_cormat <- melt(cormat) %>% rename("Trait" = Var1) %>% left_join(., sample.sizes, by = "Trait") %>% select(Trait, Var2, value, Cohort) %>% rename("Trait1" = Trait, "Cohort1" = Cohort) %>% rename('Trait' = Var2) %>% left_join(., sample.sizes, by = "Trait") %>% rename("Trait2" = Trait, "Cohort2" = Cohort)
melted_cormat$Trait1 <- factor(melted_cormat$Trait1, levels = colnames(cormat))
melted_cormat$Trait2 <- factor(melted_cormat$Trait2, levels = colnames(cormat))
one <- melted_cormat %>% select(Trait1, Cohort1) %>% set_colnames(c("Trait", "Cohort"))
two <- melted_cormat %>% select(Trait2, Cohort2)%>% set_colnames(c("Trait", "Cohort"))
rbind(one, two) %>% distinct()
col_list <- melted_cormat %>% select(Trait1, Cohort1) %>% rename("Cohort" = Cohort1) %>% merge(.,colors, by = "Cohort") %>% mutate("Trait1" = factor(Trait1, levels = colnames(cormat))) %>% distinct() %>% arrange(Trait1) %>% print()
  library(ggplot2)
  ggplot(data = melted_cormat, aes(x=Trait1, y=Trait2, fill=value)) + 
    geom_tile() +scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                                     midpoint = 0, limit = c(-1,1)) + theme_minimal(15) + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, color = col_list$color),
          axis.text.y = element_text(color = col_list$color)) + xlab("") + ylab("")
  
  #oither version, based on heatmap's hclust:
  
  cormat <- reorder_cormat(cor(n_scaled_fix[,-24]), type = "heatmap.default")
  melted_cormat <- melt(cormat) %>% rename("Trait" = Var1) %>% left_join(., sample.sizes, by = "Trait") %>% select(Trait, Var2, value, Cohort) %>% rename("Trait1" = Trait, "Cohort1" = Cohort) %>% rename('Trait' = Var2) %>% left_join(., sample.sizes, by = "Trait") %>% rename("Trait2" = Trait, "Cohort2" = Cohort)
melted_cormat$Trait1 <- factor(melted_cormat$Trait1, levels = colnames(cormat))
melted_cormat$Trait2 <- factor(melted_cormat$Trait2, levels = colnames(cormat))
one <- melted_cormat %>% select(Trait1, Cohort1) %>% set_colnames(c("Trait", "Cohort"))
two <- melted_cormat %>% select(Trait2, Cohort2)%>% set_colnames(c("Trait", "Cohort"))
rbind(one, two) %>% distinct()
col_list <- melted_cormat %>% select(Trait1, Cohort1) %>% rename("Cohort" = Cohort1) %>% merge(.,colors, by = "Cohort") %>% mutate("Trait1" = factor(Trait1, levels = colnames(cormat))) %>% distinct() %>% arrange(Trait1) %>% print()
  
  library(ggplot2)
  ggplot(data = melted_cormat, aes(x=Trait1, y=Trait2, fill=value)) + 
    geom_tile() +scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                                     midpoint = 0, limit = c(-1,1)) + theme_minimal(15) + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, color = col_list$color),
          axis.text.y = element_text(color = col_list$color)) + xlab("") + ylab("")
  
```
Well these are quite different. Maybe we should do a really good hclust before all this shiz.

```{r}
correlationHeatmapFullData <- function(din, abs = TRUE, title = "Correlation heatmap ordered by unsigned euclidian distance.")
{
  if(abs){
     distmat <- dist(t(abs(din)))
  } else
  { distmat <- dist(t((din)))}
 
o <- hclust(distmat)$order
cormat <- cor(din)
 melted_cormat <- melt(cormat) %>% rename("Trait" = Var1) %>% left_join(., sample.sizes, by = "Trait") %>% select(Trait, Var2, value, Cohort) %>% rename("Trait1" = Trait, "Cohort1" = Cohort) %>% rename('Trait' = Var2) %>% left_join(., sample.sizes, by = "Trait") %>% rename("Trait2" = Trait, "Cohort2" = Cohort)
melted_cormat$Trait1 <- factor(melted_cormat$Trait1, levels = colnames(cormat)[o])
melted_cormat$Trait2 <- factor(melted_cormat$Trait2, levels = colnames(cormat)[o])
#one <- melted_cormat %>% select(Trait1, Cohort1) %>% set_colnames(c("Trait", "Cohort"))
#two <- melted_cormat %>% select(Trait2, Cohort2)%>% set_colnames(c("Trait", "Cohort"))
#rbind(one, two) %>% distinct()
col_list <- melted_cormat %>% select(Trait1, Cohort1) %>% rename("Cohort" = Cohort1) %>% merge(.,colors, by = "Cohort") %>% mutate("Trait1" = factor(Trait1, levels = colnames(cormat)[o])) %>% distinct() %>% arrange(Trait1)
  
  library(ggplot2)
  ggplot(data = melted_cormat, aes(x=Trait1, y=Trait2, fill=value)) + 
    geom_tile() +scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                                     midpoint = 0, limit = c(-1,1)) + theme_minimal(15) + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, color = col_list$color),
          axis.text.y = element_text(color = col_list$color)) + xlab("") + ylab("") + ggtitle(title)
  col_list
}

```
Now run it...
```{r}
correlationHeatmapFullData(n_scaled_fix[,-24], abs = TRUE, title ="Correlation heatmap ordered by unsigned euclidian distance." )

correlationHeatmapFullData(n_scaled_fix[,-24], abs = FALSE, title ="Correlation heatmap ordered by signed euclidian distance." )
```

Alright. Well from these alone its hard to determine effects of cohort overlap, especially since my estimates may not be correct. Certainly there is osme of that, but also we see evidence of vertical plieotropy.

Is this effect reduced or magnified in the standard data? I expect no difference...

```{r}
correlationHeatmapFullData(z[,-24], abs = TRUE, title ="Correlation heatmap ordered by unsigned z-scores" )

correlationHeatmapFullData(z[,-24], abs = FALSE, title ="Correlation heatmap ordered by signed z-scores" )
```
 Of course this isn't going to inform me on what I want... we need an experiment
 How about this- look at stroke results and others from the UKBB. Do they cluster more closely with UKBB or their actual trait friends?
```{bash, eval = FALSE, echo = FALSE}
cd /work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/data
#Stroke:
wget https://broad-ukb-sumstats-us-east-1.s3.amazonaws.com/round2/additive-tsvs/6150_3.gwas.imputed_v3.both_sexes.tsv.bgz -O 6150_3.gwas.imputed_v3.both_sexes.tsv.bgz

#HDL:
wget https://broad-ukb-sumstats-us-east-1.s3.amazonaws.com/round2/additive-tsvs/30760_irnt.gwas.imputed_v3.both_sexes.varorder.tsv.bgz -O 30760_irnt.gwas.imputed_v3.both_sexes.varorder.tsv.bgz

#LDL:
wget https://broad-ukb-sumstats-us-east-1.s3.amazonaws.com/round2/additive-tsvs/30780_irnt.gwas.imputed_v3.both_sexes.varorder.tsv.bgz -O 30780_irnt.gwas.imputed_v3.both_sexes.varorder.tsv.bgz

#Cholesterol:
wget https://broad-ukb-sumstats-us-east-1.s3.amazonaws.com/round2/additive-tsvs/30690_irnt.gwas.imputed_v3.both_sexes.varorder.tsv.bgz -O 30690_irnt.gwas.imputed_v3.both_sexes.varorder.tsv.bgz

```
 
Now, the overall procedure to get the SNPs out from the UKBB versions.
(5/5- want to check the MAF....)
```{bash, eval = FALSE, echo = FALSE}
ls data/*.gwas.imputed_v3.both_sexes*.tsv.bgz > data/ukbb_test.traits.txt
for f in `cat data/ukbb_test.traits.txt`; do
filename=`basename $f`
OUT="${filename%.*}"
echo $OUT
#cat <(echo -e "SNP\teffect\tother\teffect_size\tstd_error\tpval") <(awk '(FNR == NR) {arr[$1]=$2;next} ($1 in arr) {print arr[$1]"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6}' lookup_snps_hg37.format.txt  <(zcat $f | tr "\t" ":" | awk -F ":" '{print $1":"$2"\t"$4"\t"$3"\t"$11"\t"$12"\t"$14}')) > analysis_data/$OUT

cat <(echo -e "SNP\teffect\tother\teffect_size\tstd_error\tpval\tMAF") <(awk '(FNR == NR) {arr[$1]=$2;next} ($1 in arr) {print arr[$1]"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7}' lookup_snps_hg37.format.txt  <(zcat $f | tr "\t" ":" | awk -F ":" '{print $1":"$2"\t"$4"\t"$3"\t"$11"\t"$12"\t"$14"\t"$6}')) > analysis_data/$OUT
done

#SPECIAL CASE because stroke is not continusou:
f="data/6150_3.gwas.imputed_v3.both_sexes.tsv.bgz"
filename=`basename $f`
OUT="${filename%.*}"
echo $OUT
cat <(echo -e "SNP\teffect\tother\teffect_size\tstd_error\tpval\tMAF") <(awk '(FNR == NR) {arr[$1]=$2;next} ($1 in arr) {print arr[$1]"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7}' lookup_snps_hg37.format.txt  <(zcat $f | tr "\t" ":" | awk -F ":" '{print $1":"$2"\t"$4"\t"$3"\t"$12"\t"$13"\t"$15"\t"$6}')) > analysis_data/$OUT
```
(note: checkd on MAF, appears correct.)
Cool, easy peasy in fact.
Now then, pull it in, harmonize and add it to matrix...
(copied from repro notebook)
```{r}
getwd()
var.list.true <- fread("/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/variant_lists/udler_final_list.csv",sep = "," )
std <- var.list.true %>% arrange(Variant) %>% select(Position_alleles, `Risk allele`, Variant) %>% rename("effect" =`Risk allele`, "SNP" = Variant) %>% separate(Position_alleles,into = c("chr", "pos", "a1", "a2"), sep = "_") 
#For some reason, this is missing one risk allele- not clear if its its T or G, we should just pick one to be consistent
std[50,]$effect = "T"
std <- std %>% mutate("other" = ifelse(a1 == effect, a2, a1))  %>% select(SNP, effect, other)
std

```
Harmonizing step...
```{r}
source("/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/harmonizeGWAS.R")
file.list <- scan("data/ukbb_test.traits.txt", what= character()) %>% gsub(x=., pattern = "data", replacement = "analysis_data") %>%
  gsub(x=., pattern = ".bgz", replacement = "")
all.dat <- lapply(file.list, function(x) fread(x) %>% mutate("effect" = toupper(effect), "other" = toupper(other)))

for(i in 1:length(all.dat))
{
  curr =  all.dat[[i]] %>% arrange(SNP) %>% select(SNP, effect, other)
  if((dim(curr)[1] == dim(std)[1])){
    print("Dimensions correct...")
    if(all(curr == std)){
    print("perfectly aligned match")
    } else{
    print("Misaligned directions:")
    all.dat[[i]] <- harmonizeEasyEffectOnly(all.dat[[i]], std)
    }
  } else{
    print("Issues for:")
    print(i)
    all.dat[[i]] <- harmonizeEasyEffectOnly(all.dat[[i]], std)
  }
  
}
```
Okay. Phew. Now can we insert? Order correct?
```{r}
stopifnot(all(all.dat[[1]]$SNP == udler_Beta$SNP))
joined <- cbind(z[,-24], data.frame(sapply(all.dat, function(x) x$effect_size/x$std_error)) %>% set_colnames(c("cholesterol_2", "HDL_2", "LDL_2", "Stroke_2")))
correlationHeatmapFullData(joined, abs = FALSE, title ="Correlation heatmap clustered by signed z_scores." )
#order seems correct....

correlationHeatmapFullData(joined, abs = TRUE, title ="Correlation heatmap ordered by unsigned z_scores." )
```
Well based on this, the cohort-specific effect is pretty minor. What about in PCA?
```{r}
pca <- svd(joined,nv = 10)
#get teh colors we like....
ct <- sample.sizes %>% left_join(., colors, by = "Cohort") %>% filter(Trait %in% colnames(joined)) %>% mutate("Trait" = factor(Trait, levels = colnames(joined))) %>% arrange(Trait)
plotFactors(pca$v, trait_names = colnames(joined), title = "", colors = ct$color)
```
Grouping on stroke
```{r}
ct$shape <- ifelse(grepl(pattern = "Stroke", x = ct$Trait),22,19)
pca2 <- prcomp(t(joined),rank. = 10,center = TRUE, scale. = TRUE)
plot(pca2$x[,1], pca2$x[,2], col=ct$color, pch = ct$shape)

plot(pca2$x[,2], pca2$x[,3], col = ct$color, pch = ct$shape)
```
Issue with this test- only across 94 SNPs that are significant for T2D. So might just be the strong signal here....
Test on the LDL/triglcerides ones
```{r}
ct$shape2 = ifelse(grepl(pattern = "cholesterol", x = ct$Trait),19,1)
ct$shape2[(ct$Trait == "hdl" | ct$Trait == "HDL_2")] <- 15
ct$shape2[(ct$Trait == "ldl" | ct$Trait == "LDL_2")] <- 17

plot(pca2$x[,1], pca2$x[,2], col=ct$color, pch = ct$shape2, xlab = "PC1", ylab = "PC2")

plot(pca2$x[,2], pca2$x[,3], col = ct$color, pch = ct$shape2)
```
TODO: try with more snps.s

#5/05
Try in flashr
```{r}
original.flash <- flashier::flash(as.matrix(z[,-24]))
with.added.flash <- flashier::flash(as.matrix(joined))
```
Visualize- I want a barplot colored by cohort.
```{r}
col_list <- melted_cormat %>% select(Trait1, Cohort1) %>% rename("Cohort" = Cohort1) %>% merge(.,colors, by = "Cohort")  %>% distinct() %>% arrange(Trait1) %>% print()


ordered.colors <- col_list %>% mutate("traits.ordered" = factor(Trait1, levels = colnames(z[,-24]))) %>% arrange(traits.ordered)
plotFactorsBarplot(original.flash$loadings.pm[[2]], colnames(z[,-24]), "Colors",colors = ordered.colors$Cohort)

plotFactors(original.flash$loadings.pm[[2]], trait_names = colnames(z[,-24]), title = "", colors = ordered.colors$color)
```

Now adding in the other traits. Do they group by trait or by cohort?
```{r}
#ordered.colors <- col_list %>% mutate("traits.ordered" = factor(Trait1, levels = colnames(joined))) %>% arrange(traits.ordered)
col_list$Trait1 <- as.character(col_list$Trait1)
col_list.joined <- rbind(col_list, c("UKBB", "cholesterol_2","#1F78B4"), c("UKBB", "HDL_2","#1F78B4"), c("UKBB", "LDL_2","#1F78B4"), c("UKBB", "Stroke_2","#1F78B4"))
ordered.colors <- col_list.joined  %>% mutate("traits.ordered" = factor(Trait1, levels = colnames(joined))) %>% arrange(traits.ordered)
plotFactorsBarplot(with.added.flash$loadings.pm[[2]], colnames(joined), "With added UKBB traits", colors = ordered.colors$Cohort)

plotFactors(with.added.flash$loadings.pm[[2]], trait_names = colnames(joined), title = "With added UKBB traits", colors = ordered.colors$color)
```

I guess the real question is how does it cluster them?
```{r}
K =length(unique(col_list$Cohort))
small.group <- hclust(dist(original.flash$loadings.pm[[2]]))
plot(small.group)
clust_assigns <- cutree(small.group, K)
hclust.intuit <- data.frame("Trait1" = names(clust_assigns), "Group" = clust_assigns) %>% merge(col_list, by = "Trait1")
ggplot(hclust.intuit, aes(x = as.factor(Group), y = Cohort, color = Cohort, label = as.character(Trait1)))  + geom_jitter(position = position_jitter(seed = 1)) + geom_text(position = position_jitter(seed = 1), angle = 45, hjust = 1, vjust = -1) + theme_minimal(15) + xlab("Clustered group") 


ggplot(hclust.intuit, aes(x = as.factor(Group), y = Cohort, color = Cohort, label = as.character(Trait1)))  + geom_jitter(width = 0.2, height = 0.3)  + theme_minimal(15) + xlab("Clustered group") 

```
Now, with the added in thing....
```{r}
K =length(unique(col_list.joined$Cohort))
full.group <- hclust(dist(with.added.flash$loadings.pm[[2]]))
plot(full.group)
clust_assigns <- cutree(full.group, K)
relevant_labels <- c("ldl", "LDL_2", "Stroke_2", "HDL_2", "Stroke_LVD", "total_cholesterol", "hdl", "cholesterol_2", "Stroke_SVD")
hclust.intuit <- data.frame("Trait1" = names(clust_assigns), "Group" = clust_assigns) %>% merge(col_list.joined, by = "Trait1") %>% mutate("labs" = ifelse(as.character(Trait1) %in% relevant_labels, as.character(Trait1), ""))


ggplot(hclust.intuit, aes(x = as.factor(Group), y = Cohort, color = Cohort, label = as.character(labs)))  + geom_jitter(position = position_jitter(seed = 1)) + geom_text(position = position_jitter(seed = 1), angle = 45, hjust = 1, vjust = -1) + theme_minimal(15) + xlab("Clustered group") 


ggplot(hclust.intuit, aes(x = Cohort, y = as.factor(Group), color = Cohort, label = as.character(labs)))  + geom_jitter(position = position_jitter(seed = 1,height = 0.2)) + geom_text(position = position_jitter(seed = 1, height = 0.3),hjust = 0.5, vjust = -0.5)  + xlab("GWAS Cohort") + theme_minimal(15) + theme(panel.grid.major.x = element_blank(), axis.text.x=element_text(angle = 45, hjust = 0.5),panel.grid.minor.y = element_line()) + ylab("Clustered group")


ggplot(hclust.intuit, aes(x = as.factor(Group), y = Cohort, color = Cohort, label = as.character(Trait1)))  + geom_jitter(width = 0.2, height = 0.3)  + theme_minimal(15) + xlab("Clustered group") 
```


Sanity check: are the z score the same direction for these 2?
```{r}
joined %>% select("LDL_2", "ldl") %>% mutate("same_dir" = ifelse(sign(LDL_2) == sign(ldl), TRUE, FALSE)) %>% group_by(same_dir) %>% summarize("count" = n())

joined %>% select("HDL_2", "hdl") %>% mutate("same_dir" = ifelse(sign(HDL_2) == sign(hdl), TRUE, FALSE)) %>% group_by(same_dir) %>% summarize("count" = n())

joined %>% select("Stroke_2", "Ischemic_Stroke", "Stroke_SVD") %>% mutate("same_dir" = ifelse(sign(Stroke_2) == sign(Ischemic_Stroke) | sign(Stroke_2) == sign(Stroke_SVD), TRUE, FALSE)) %>% group_by(same_dir) %>% summarize("count" = n())
```
Well, that's not good. Large proportions of these don't even go the same direction. Did I mix up the traits?
Checking this. Looks like the numbers are associated correctly..
Are the MAFs way different?
```{r}
/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/data
```


How about PVE? Do these differ on just the first significant ones?

```{r}
K =length(unique(col_list.joined$Cohort))
sig_factors <- sum(with.added.flash$pve > (1/with.added.flash$n.factors))
full.group <- hclust(dist(with.added.flash$loadings.pm[[2]][,1:3]))
plot(full.group)
clust_assigns <- cutree(full.group, K)
relevant_labels <- c("ldl", "LDL_2", "Stroke_2", "HDL_2", "Stroke_LVD", "total_cholesterol", "hdl", "cholesterol_2", "Stroke_SVD")
hclust.intuit <- data.frame("Trait1" = names(clust_assigns), "Group" = clust_assigns) %>% merge(col_list.joined, by = "Trait1") %>% mutate("labs" = ifelse(as.character(Trait1) %in% relevant_labels, as.character(Trait1), ""))

ggplot(hclust.intuit, aes(x = Cohort, y = as.factor(Group), color = Cohort, label = as.character(labs)))  + geom_jitter(position = position_jitter(seed = 1,height = 0.2)) + geom_text(position = position_jitter(seed = 1, height = 0.3),hjust = 0.5, vjust = -0.5)  + xlab("GWAS Cohort") + theme_minimal(15) + theme(panel.grid.major.x = element_blank(), axis.text.x=element_text(angle = 45, hjust = 0.5),panel.grid.minor.y = element_line()) + ylab("Clustered group")

```


One option is to test the residuals, eh?
"What i should have done from the get-go...."
As in, are the residuals for a certain cohort random, or do they have some kind of structure?

```{r}
with.added.flash$loadings.pm[[1]]

#testing L"
lm(as.matrix(joined[,..col]) ~ as.matrix(with.added.flash$loadings.pm[[1]]))
   #testing F
summary.f <- lapply(1:nrow(joined), function(row) lm(as.matrix(unlist(joined[row,])) ~ as.matrix(with.added.flash$loadings.pm[[2]])))
residuals.f <- sapply(summary.f, function(x) resid(x))
fitted.f <- sapply(summary.f, function(x) fitted(x))
plot(as.vector(residuals.f),as.vector(fitted.f))
```
Looks like appretty decent fit to me. I think we would need to color by cohort though to be sure....
Or do by cohort, huh
```{r}
cohorts <- unique(col_list.joined$Cohort)
for(c in cohorts)
{
ukbb <- (filter(col_list.joined, Cohort == c))$Trait1
 print(plot(as.vector(residuals.f[ukbb,]),as.vector(fitted.f[ukbb,]), main = c))
}
```

This overall looks pretty good to me. I think in a case like this, maybe we don't need to correct?
then again, maybe for some of these.... get a cook's stat or something?




Note- good chat with surya, diptavo, a few thigns to try:
1) can I really amplify that R2 effect? What makes it go? what can I do to amp it up? Use that simulation.
With surya- we want ot scale by root N for sure
Reasoning?
#5/18
Try the real deal whitening.
```{r}
ldsc.r2g.data <- list.files("/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/ldsr_all/ldsr_results",pattern = "*rg_report.tsv")
inter <- lapply(ldsc.r2g.data, function(x) fread(paste0("/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/ldsr_all/ldsr_results/", x)) %>% mutate("Source" = x))
joined.main <- do.call("rbind", inter)
```
With this, I need to make a matrix that has the `gcov_int` as an entry for every p1/p2 combo.
We have done a terrible job keeping the naming consistent, so the whole thing is straight up mangled.
Really sucks.

I got fed up with all this and just typed out all the names of the input files and the traits they coorrespond to.
On my desktop as thenamesofthedead.
this should help out with this crab.
```{r}
overlap.mat <- joined.main %>% select(p1, p2, gcov_int) %>% mutate("id" = paste0(p1, ":", p2)) %>% distinct_at(vars(id), .keep_all = TRUE) %>% select(-id) %>% pivot_wider(values_from = gcov_int, names_from =p1) %>% arrange(p2) %>% print()

#which ons are missing?
colnames(overlap.mat)[!(colnames(overlap.mat) %in% overlap.mat$p2)]
overlap.mat$p2[!(overlap.mat$p2 %in% colnames(overlap.mat))]
#Squareify it.
#Simple approach
matrix.fetch.you <- as.matrix(overlap.mat[,-1])
rownames(matrix.fetch.you) <- overlap.mat$p2
colnames(matrix.fetch.you) <- colnames(overlap.mat)[-1]
name.order <- colnames(matrix.fetch.you)
matrix.fetch.you.correct <- matrix.fetch.you[name.order, name.order]
#Add on the missing dumb column
missing.entries <- which(!(overlap.mat$p2 %in% colnames(overlap.mat)))
for(m in missing.entries)
{
  oldnames <- colnames(matrix.fetch.you.correct)
  matrix.fetch.you.correct <- rbind(matrix.fetch.you.correct, matrix.fetch.you[m,name.order] )
  rownames(matrix.fetch.you.correct) <- c(oldnames, overlap.mat$p2[m])
  matrix.fetch.you.correct <- cbind(matrix.fetch.you.correct, c(matrix.fetch.you[m,name.order], 1))
  colnames(matrix.fetch.you.correct) <- c(oldnames, overlap.mat$p2[m])
}
diag(matrix.fetch.you.correct) <- 1
#fillifyMatrix(matrix.fetch.you.correct) 
image(matrix.fetch.you.correct)
isSymmetric(matrix.fetch.you.correct)
#need to check symmetry ast this point. urg.
```
Wwe have an issue here- 2 of these have a correlation greater than one, and large sample size and I think are highly heritable.
The exact condition where we expect inflation to occur. Poop.
To adjust, we follow the guideline in https://www.biorxiv.org/content/10.1101/310565v1.full.pdf, which accounts for Fst2, population size and correlation estimate:
(traits are LDL and Cholesterol from the UKBB.
/work-zfs/abattle4/lab_data/UKBB/GWAS_Neale/ldsr_format/30690_irnt.imputed_v3.ldsc.both_sexes.tsv.gz
/work-zfs/abattle4/lab_data/UKBB/GWAS_Neale/ldsr_format/30780_irnt.imputed_v3.ldsc.both_sexes.tsv.gz
```{r}
rg.mat <- ldscTableToMatrix(joined.main, "rg")
ps <- unlist(rg.mat[4,6])
Fst <- 0.001 #from https://humgenomics.biomedcentral.com/articles/10.1186/s40246-022-00380-5#Sec10, the min.
N1 <- 343621
N2 <-344278
adjusted.gcovint <- matrix.fetch.you.correct[4,6] - (ps * Fst^2 * sqrt(N1 * N2))
matrix.fetch.you.correct[4,6] <- adjusted.gcovint
matrix.fetch.you.correct[6,4] <- adjusted.gcovint

```


I think this is right!
Let's check and then fix the diags
```{r}
stopifnot(all(colnames(new.overlap.mat)[-1] == new.overlap.mat$p2))
ldscintmatrix <- matrix.fetch.you.correct
```

Now, I want to just get the names right
```{r}
curr.names <- gsub(colnames(ldscintmatrix), pattern = "/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/ldsr_all/munged_sumstats/", replacement = "")
new.names <- fread("/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/files_to_traits.ldsc.csv", fill=TRUE, header =  TRUE) 
head(new.names)
if(all(curr.names == new.names$LDSCPath)) {
  colnames(ldscintmatrix) <- new.names$Trait
  rownames(ldscintmatrix) <- new.names$Trait
}
ldscintmatrix[is.na(ldscintmatrix)] <- 0

#set up the colors
new.colors.joined <- left_join(new.names, colors, by = "Cohort")
stopifnot(all(new.colors.joined$LDSCPath == new.names$LDSCPath))
plotCorrelationHeatmap(ldscintmatrix,colors = new.colors.joined, typin = "heatmap.default")
```
Phew. I think we are there!
Also need the r2g to compare....
```{r}
rg.mat <- ldscTableToMatrix(joined.main, "rg")
#Name check
curr.names <- gsub(colnames(rg.mat), pattern = "/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/ldsr_all/munged_sumstats/", replacement = "")
if(all(curr.names == new.names$LDSCPath)) {
  colnames(rg.mat) <- new.names$Trait
  rownames(rg.mat) <- new.names$Trait
}
plotCorrelationHeatmap(rg.mat,colors = new.colors.joined, typin = "heatmap.default")

```
I'm thinking that the pairs of traits with really low h2 estimates are the ones that are problmeatic (r2 estimates that are too high)
In the case of na, those are for sure those with low h2, or negative h2
```{r}
problematic.too.high <- which(apply(rg.mat, 2, function(x) sum(x > 1)) != 0)
problematic.too.high.pairwise <- apply(rg.mat, 1, function(x) which(x > 1))

problematic.na <- which(rowSums(rg.mat) == 1)
length(unique(joined.main$p2))
ordered.by.h2 <- joined.main %>% group_by(p2) %>% slice(which.min(h2_obs_se)) %>%  ungroup() %>% arrange(h2_obs) %>% select(p2, h2_obs, h2_obs_se) %>% 
  mutate("LDSCPath" = gsub(x = p2, pattern = "/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/ldsr_all/munged_sumstats/", replacement = "")) %>%
  left_join(., new.names, by = "LDSCPath") %>% select(-p2, -LDSCPath) %>% print()
#problematic.na
names(problematic.too.high)
new.names

```
We can see that those with low heritability estimates usually aren't able to get good correlation estimates.
Those with reall
```{r}
df.plot.h2 <- ordered.by.h2 %>% mutate("coloration1" = ifelse(Trait %in% names(problematic.too.high), "False", "True"), "coloration2" = ifelse(Trait %in% names(problematic.na), "False", "True"))
ggplot(df.plot.h2, aes(x = reorder(Trait, h2_obs), y = h2_obs, color = coloration1)) + geom_point() + 
  xlab("Trait") + ylab(expression(h^2)) + theme_classic() + ggtitle("Studies with irregular correlation estimates") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
ggplot(df.plot.h2, aes(x = reorder(Trait, h2_obs), y = h2_obs, color = coloration2)) + geom_point() +   xlab("Trait") + ylab(expression(h^2)) + theme_classic() + ggtitle("Studies lacking correlation estimates") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + labs(color = "rg estimate") 


ggplot(df.plot.h2, aes(x = h2_obs_se, y = h2_obs, color = coloration1)) + geom_point() + 
  xlab("h2_se") + ylab(expression(h^2)) + theme_classic() + ggtitle("Studies with irregular correlation estimates")
ggplot(df.plot.h2, aes(x = h2_obs_se, y = h2_obs, color = coloration2)) + geom_point() + 
  xlab("h2_se") + ylab(expression(h^2)) + theme_classic() + ggtitle("Studies lacking correlation estimates")

```
Need to look at them pairwise too....
Why are these traits so sucky?
```{r}
problematic.too.high.pairwise <- apply(rg.mat, 1, function(x) which(x > 1))
pairwise <- NULL
for(i in 1:length(problematic.too.high.pairwise))
{
  n <- names(problematic.too.high.pairwise)[i]
  if(length(problematic.too.high.pairwise[[i]]) > 0)
  {
    for(entry in names(problematic.too.high.pairwise[[i]]))
    {
      pairwise <- rbind(pairwise, c(n, entry))
    }
  }
}
pairwise.ids <- data.frame(pairwise) %>% mutate("uid" = paste0(X1, ":", X2))
#get the data for these...
pairwise.errors <- joined.main %>% mutate("LDSCPath" = gsub(x = p2, pattern = "/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/ldsr_all/munged_sumstats/", replacement = "")) %>%  left_join(., new.names, by = "LDSCPath") %>% rename("Trait1" = Trait) %>% select(-LDSCPath) %>% mutate("LDSCPath" = gsub(x = p1, pattern = "/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/ldsr_all/munged_sumstats/", replacement = "")) %>%  left_join(., new.names, by = "LDSCPath") %>% rename("Trait2" = Trait) %>% select(Trait1, Trait2, rg, se, z, p, h2_obs,h2_obs_se,h2_int, h2_int_se,gcov_int,gcov_int_se) %>% mutate("uid" = paste0(Trait1, ":", Trait2)) %>% mutate("coloration" = ifelse(uid %in% pairwise.ids$uid, "ingroup", "regular")) %>% drop_na()
pairwise.errors

ggplot(pairwise.errors, aes(x = h2_obs_se, y = rg, color = coloration)) + geom_point() + 
  xlab("h2_se") + ylab("Rg") + theme_classic() + ggtitle("Studies with correlation estimates > 1")

ggplot(pairwise.errors, aes(x = h2_obs, y = rg, color = coloration)) + geom_point() + 
  xlab(expression(h^2)) + ylab("Rg") + theme_classic()  + ggtitle("Studies with correlation estimates > 1")

ggplot(pairwise.errors, aes(x = gcov_int, y = rg, color = coloration)) + geom_point() + 
  xlab("gcov intercept") + ylab("Rg") + theme_classic()  + ggtitle("Studies with correlation estimates > 1")


ggplot(pairwise.errors, aes(x = gcov_int_se, y = rg, color = coloration)) + geom_point() + 
  xlab("gcov intercept se") + ylab("Rg") + theme_classic()  + ggtitle("Studies with correlation estimates > 1")
```

With this kind of done, we want to see how well the intrecept group actually groups by known cohort overlap, and identifies potential new cohort overlap...
Following the same procedure as in `fastAsset`
```{r}
#ldscintmatrix[ldscintmatrix > 1] <- 1 #for these, need to adjust with FST, rg, and sample size...
#ldscintmatrix[ldscintmatrix < -1] <- -1
if(!isSymmetric(ldscintmatrix))
{
  ldscintmatrix <- fullifyMatrix(ldscintmatrix)
}
names_order <- colnames(ldscintmatrix)
#ensure diagonals are 1
ldscintmatrix = solve(diag(sqrt(diag(ldscintmatrix)))) %*% ldscintmatrix %*% solve(diag(sqrt(diag(ldscintmatrix))))
rownames(ldscintmatrix) <- names_order
colnames(ldscintmatrix) <- names_order
# Hierarchical clustering
corrdist = as.dist(1-abs(ldscintmatrix))
hc = hclust(corrdist)
library(dendextend)
dend <- as.dendrogram(hc)
labels_colors(dend) <- new.colors.joined$color[order.dendrogram(dend)]

plot(dend)
```

We have a few different options to see what our tree looks like...
```{r}
#few different options- could just go by cohort, and see what we get...

htree = cutree(hc, h=0.8) # Criterion: corr>0.2
block = as.integer(names(table(htree))[table(htree)>=2]) #only include clusters with at least 2 traits in them
block = lapply(block, function(x) names(htree)[htree==x]) #get the corresponding names
```

With this made, we only do correction within those groups.
Are the groups of concern pupt together?
Now, I need to just get the results for those blocks, and clear out all else.
```{r}
traits_i <- colnames(ldscintmatrix)
  #already done above    
#traits_i_block = sapply(block, function(x) traits_i[traits_i%in%x])
    #traits_i_block = block[sapply(block,length)>=2]
  #Guanghao then corrects block by block, since the rest shouldn't be touched, huh...
o <- order(block[[1]])
#For each block, whiten the snps.
ldscint.whitening <- diag(1, nrow = nrow(ldscintmatrix) )
rownames(ldscint.whitening) <- rownames(ldscintmatrix)
colnames(ldscint.whitening) <- colnames(ldscintmatrix)
for(b in block)
{
  ldscint.whitening[b,b] <- ldscintmatrix[b,b]
}
plotCorrelationHeatmap(ldscint.whitening, colors = NA, title = "Matrix for whitening")
```
Get the orders and everythign updated
```{r}
data.order <-  fread("/work-zfs/abattle4/ashton/snp_networks/scratch/udler_td2/ldsr_all/ldsc_name_mapping_again.csv", header = FALSE)
o <- rownames(with.added.flash$loadings.pm[[2]])
data.order$V1 <- factor(data.order$V1, levels = o)
data.order <- data.order %>% filter(V1 != "") %>% arrange(V1)
final.selected.whitening <- ldscint.whitening[data.order$V2,data.order$V2 ]
```


Noe doing whitening like guanhao does....
```{r, echo = FALSE, eval = FALSE}
t <- svd(final.selected.whitening)
Ui <- solve(chol(final.selected.whitening))
z.o <- t(as.matrix(Ui) %*% t(as.matrix(joined)))
image(as.matrix(joined))
image(z.o)
```
Now then, let's go ahead and do the test weve been waiting for....


```{r}
whitened.flash <- flashier::flash(as.matrix(z.o))
```
Factor plot
```{r}
plotFactors(whitened.flash$loadings.pm[[2]], trait_names = rownames(whitened.flash$loadings.pm[[2]]), title = "", colors = ordered.colors$color)
```


```{r}
K =length(unique(col_list.joined$Cohort))
full.group <- hclust(dist(whitened.flash$loadings.pm[[2]]))
plot(full.group)
clust_assigns <- cutree(full.group, K)
relevant_labels <- c("LDL", "LDL_UKBB", "Stroke_UKBB", "HDL_UKBB", "Stroke_LVD", "Cholesterol_UKBB", "HDL", "total_cholesterol", "Stroke_SVD")
hclust.intuit <- data.frame("Trait" = names(clust_assigns), "Group" = clust_assigns) %>% merge(new.colors.joined, by = "Trait") %>% select(-LDSCPath) %>% mutate("labs" = ifelse(as.character(Trait) %in% relevant_labels, as.character(Trait), ""))

ggplot(hclust.intuit, aes(x = Cohort, y = as.factor(Group), color = Cohort, label = as.character(labs)))  + geom_jitter(position = position_jitter(seed = 1,height = 0.2)) + geom_text(position = position_jitter(seed = 1, height = 0.3),hjust = 0.5, vjust = -0.5)  + xlab("GWAS Cohort") + theme_minimal(15) + theme(panel.grid.major.x = element_blank(), axis.text.x=element_text(angle = 45, hjust = 0.5),panel.grid.minor.y = element_line()) + ylab("Clustered group")
```
Well there ya go. Whitening does not appear to fix the fetching problem, huh.
Try with correfcting for N, try on the hm3 dataset.

#still stinky code
That was excruciating. Now I can generate the whitenin gmatrix
```{r}
#https://github.com/gqi/fastASSET/blob/main/R/fastASSET.R
overlap.mat #we need to get this in the same order as everything
ordered <- if(overlap.mat$p2 %in% colnames(overlap.mat)
              colnames(overlap.mat) %in% overlap.mat$p2 



#For NA entries, we make them 0:
ldscint.mat <- as.matrix(overlap.mat[,-1])
ldscint.mat[is.na(ldscint.mat)] <- 0
```
CODE from fastASset
```{r}
#ensure diagonals are 1
ldscintmat = solve(diag(sqrt(diag(ldscint.mat)))) %*% ldscint.mat %*% solve(diag(sqrt(diag(ldscint.mat))))
colnames(ldscintmat) = rownames(ldscintmat) = traits
# Hierarchical clustering
corrdist = as.dist(1-abs(ldscintmat))
hc = hclust(corrdist)
htree = cutree(hc, h=0.8) # Criterion: corr>0.2
block = as.integer(names(table(htree))[table(htree)>=2])
block = lapply(block, function(x) names(htree)[htree==x])


corrected.z.scores <- lapply(1:nrow(all.combs), function(j) {i=all.combs[j,]; t(solve(t(chol(whitener.matrices[[j]]))) %*%
                                                         t(cbind(pop1.condition[[i[1]]], pop2.condition[[i[2]]])))})
```


#missing one, but tbh don't care. Just want to get this done oh my gosh.
overlap.mat <- overlap.mat %>% select(c(p2, as.character(t$raw )))
col_order <- colnames(overlap.mat)

t$raw <- factor(t$raw, levels = col_order[-1])
t <- t %>% arrange(raw)

colnames <- lapply(basename(colnames(overlap.mat)), function(x) str_split(x, pattern = "\\.")[[1]][1])
rnames <- lapply(basename(overlap.mat$p2), function(x) str_split(x, pattern = "\\.")[[1]][1])


if(colnames[2:length(colnames)] != rnames)
{
  
}
colnames[2]
colnames(overlap.mat)
mat <- matrix(NA, nrow = length(trait.list), ncol = length(trait.list))
colnames(mat) <- trait.list
rownames(mat) <- trait.list
for(t in trait.list)
{
  for(t2 in trait.list){
    i = which(trait.list == t)
    j = which(trait.list == t2)
    lookup <- overlap.mat %>% filter(p1 == t) %>% select(!!sym(t2))
    if(i ==j)
    {
      mat[i,j] <- 1
    } else if(nrow(lookup) == 0 | is.na(lookup[1,1]))
    {
      #mat[i,j] <- NA
      next
    } else
    {
      mat[i,j] <- unlist(lookup)[1]
      mat[j,i] <- unlist(lookup)[1]
    }
  }
}
write.table(data.frame(mat),file = "./ldsc_intercepts.cor.tsv",quote = FALSE,sep = '\t',row.names = FALSE, col.names = T)
fread("./ldsc_intercepts.cor.tsv")
getwd()
```

#sad code: trying to work on the list stuff to make the matrix
for(i in 2:ncol(overlap.mat))
{
  curr_frick <- colnames(overlap.mat)[i]
  names_list <- c(names_list, curr_frick)
  if(curr_frick %in% overlap.mat$p2)
  {
    new_mat[[i-1]] <- overlap.mat[which(overlap.mat$p2 == curr_frick),]
  }else{
    print("missing this...")
    print(curr_frick)
  }
}
#Make this into a matrix, then add in the missing ones as a new row and nuew column
new_mat <- list()
names_list <- c()


new.overlap.mat <- data.frame(do.call("rbind", new_mat))
colnames(new.overlap.mat) <- colnames(overlap.mat)

for (m in missing.entries)
{
  new.overlap.mat <- rbind(new.overlap.mat, overlap.mat[m,])
  str.name <- unlist(overlap.mat[m,1]$p2)
  new.overlap.mat <- cbind(new.overlap.mat, c(t(overlap.mat[m,-1]), 1))
  last.entry <- ncol(new.overlap.mat)
  colnames(new.overlap.mat)[last.entry] <- str.name
}




